source/extensions/retry/host/other_hosts/other_hosts.h

```diff
@@ -6,14 +6,17 @@
 namespace Envoy {
 class OtherHostsRetryPredicate : public Upstream::RetryHostPredicate {
 public:
+  OtherHostsRetryPredicate(uint32_t retry_count) : attempted_hosts_(retry_count) {}
+
   bool shouldSelectAnotherHost(const Upstream::Host& candidate_host) override {
-    return attempted_hosts_.find(candidate_host.address()->asString()) != attempted_hosts_.end();
+    return std::find(attempted_hosts_.begin(), attempted_hosts_.end(),
+                     candidate_host.address()->asString()) != attempted_hosts_.end();
   }
   void onHostAttempted(Upstream::HostDescriptionConstSharedPtr attempted_host) override {
-    attempted_hosts_.insert(attempted_host->address()->asString());
+    attempted_hosts_.emplace_back(attempted_host->address()->asString());
   }
 
 private:
-  std::unordered_set<std::string> attempted_hosts_;
+  std::vector<std::string> attempted_hosts_;
```

Comment:
Do we need to store a string here? Or can we just use host pointer comparisons? I suppose it's possible that hosts could be duplicated for some reason but that seems like the rare case?

Reply:
Hosts can't be duplicated by address since we key off the address during host updates (modulo eventual consistency - since we update one priority at a time I guess it's technically possible mid-update). The other case where you'd get surprising behavior would be if a host gets removed then added again.

That said, I think both of those cases would be extremely rare so I'd be okay with using pointer comparisons.